### PR TITLE
flutter/tests support

### PR DESCRIPTION
**Pull request by Hixie**
This introduces a script that implements the conventions described in https://github.com/flutter/tests/pull/1.

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -229,6 +233,7 @@ task:
     cpu: 4
   env:
     CIRRUS_WORKING_DIR: "C:\\Windows\\Temp\\flutter sdk"
+    PATH: "$CIRRUS_WORKING_DIR/bin;$CIRRUS_WORKING_DIR/bin/cache/dart-sdk/bin;$PATH"
```
Comment (tvolkert) [inline .cirrus.yml]: Aren't Windows environment variables expressed as `%ENV_VAR%`?

File: dev/customer_testing/run_tests.dart [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+Future<bool> run(List<String> arguments) async {
+  final ArgParser argParser = ArgParser(
+    allowTrailingOptions: false,
+    usageLineLength: 72,
+  )
+    ..addOption(
+      'repeat',
+      defaultsTo: '1',
+      help: 'How many times to run each test. Set to a high value to look for flakes.',
```
Comment (tvolkert) [inline dev/customer_testing/run_tests.dart]: `valueHelp: 'count'`

File: dev/customer_testing/run_tests.dart [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    )
+    ..addFlag(
+      'skip-on-fetch-failure',
+      defaultsTo: false,
+      help: 'Whether to skip tests that we fail to download.',
+    )
+    ..addFlag(
+      'skip-template',
+      defaultsTo: false,
+      help: 'Whether to skip tests named "template.dart".',
```
Comment (tvolkert) [inline dev/customer_testing/run_tests.dart]: You say "template.dart", but the code looks for "template.test"

File: dev/customer_testing/run_tests.dart [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    .toList();
+
+  if (help || repeat == null || files.isEmpty) {
+    print('run_tests.dart [options...] path/to/file1.test path/to/file2.test...');
+    print('For details on the test registry format, see:');
+    print('  https://github.com/flutter/tests/blob/master/registry/template.test');
+    print('');
+    print(argParser.usage);
+    print('');
+    return true;
```
Comment (tvolkert) [inline dev/customer_testing/run_tests.dart]: Nit: if `repeat == null`, I think false would be a more appropriate return value.

File: dev/customer_testing/run_tests.dart [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+      throw FormatException('${errorPrefix}No "test" directives specified. At least one command must be specified to run tests.');
+    return TestFile._(
+      List<String>.unmodifiable(contacts),
+      List<String>.unmodifiable(fetch),
+      List<Directory>.unmodifiable(update),
+      List<String>.unmodifiable(test),
+    );
+  }
+
+  TestFile._(this.contacts, this.fetch, this.update, this.tests);
```
Comment (tvolkert) [inline dev/customer_testing/run_tests.dart]: const

File: dev/customer_testing/run_tests.dart [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    print('$failures failure$s.');
+    return false;
+  }
+  if (verbose) {
+    print('All tests passed!');
+  }
+  return true;
+}
+
+class TestFile {
```
Comment (tvolkert) [inline dev/customer_testing/run_tests.dart]: `@immutable`

File: dev/customer_testing/run_tests.dart [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    print('>> $command');
+  Process process;
+  if (Platform.isWindows) {
+    process = await Process.start('CMD.EXE', <String>['/S', '/C', '$command'], workingDirectory: directory.path);
+  } else {
+    final List<String> segments = command.trim().split(_spaces);
+    process = await Process.start(segments.first, segments.skip(1).toList(), workingDirectory: directory.path);
+  }
+  final List<String> output = <String>[];
+  process.stdout.transform(utf8.decoder).transform(const LineSplitter()).listen(verbose ? printLog : output.add);
```
Comment (tvolkert) [inline dev/customer_testing/run_tests.dart]: `utf8.decoder.bind(process.stdout).transform(...)` - it'll make it more resilient to us changing stdout to be a stream of Uint8List down the road.

(and below)

File: .cirrus.yml (hunk 1 above)
Comment (Hixie) [reply .cirrus.yml]: I copied and pasted this from elsewhere in the same file and it fixed my problem. I've no idea what the syntax of this string is supposed to be or how it's implemented.

File: dev/customer_testing/run_tests.dart (hunk 2 above)
Comment (Hixie) [reply dev/customer_testing/run_tests.dart]: done

File: dev/customer_testing/run_tests.dart (hunk 3 above)
Comment (Hixie) [reply dev/customer_testing/run_tests.dart]: fixed

File: dev/customer_testing/run_tests.dart (hunk 4 above)
Comment (Hixie) [reply dev/customer_testing/run_tests.dart]: replaced with `return help`.

File: dev/customer_testing/run_tests.dart (hunk 5 above)
Comment (Hixie) [reply dev/customer_testing/run_tests.dart]: done

File: dev/customer_testing/run_tests.dart (hunk 6 above)
Comment (Hixie) [reply dev/customer_testing/run_tests.dart]: done, added meta, reupdated pubspecs.

File: dev/customer_testing/run_tests.dart (hunk 7 above)
Comment (Hixie) [reply dev/customer_testing/run_tests.dart]: that's much less idiomatic. Will that really be required?

Comment (Hixie) [reply dev/customer_testing/run_tests.dart]: better add these to your list...

**Review by tvolkert — APPROVED** [review]

**Pull request merged**